opentracing-jdbc/opentracing-jdbc/src/main/java/org/zalando/opentracing/jdbc/TracingQueryExecutionListener.java

```diff
@@ -0,0 +1,61 @@
+package org.zalando.opentracing.jdbc;
+
+import io.opentracing.Span;
+import io.opentracing.Tracer;
+import lombok.AllArgsConstructor;
+import lombok.SneakyThrows;
+import net.ttddyy.dsproxy.ExecutionInfo;
+import net.ttddyy.dsproxy.QueryInfo;
+import net.ttddyy.dsproxy.listener.QueryExecutionListener;
+import org.zalando.opentracing.jdbc.operation.OperationName;
+import org.zalando.opentracing.jdbc.span.SpanDecorator;
+
+import java.lang.reflect.Method;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.List;
+
+import static java.util.stream.Collectors.toList;
+
+@AllArgsConstructor
+final class TracingQueryExecutionListener implements QueryExecutionListener {
+
+    private static final String SPAN = TracingQueryExecutionListener.class.getName() + ".SPAN";
+
+    private final Tracer tracer;
+    private final OperationName operationName;
+    private final SpanDecorator decorator;
+
+    @Override
+    @SneakyThrows(SQLException.class)
+    public void beforeQuery(final ExecutionInfo info, final List<QueryInfo> infos) {
+        final Method method = info.getMethod();
+        final List<String> queries = infos.stream()
+                .map(QueryInfo::getQuery)
+                .collect(toList());
+
+        final String name = operationName.generate(method, queries);
+        final Span span = tracer.buildSpan(name).start();
+        final Statement statement = info.getStatement();
+
+        decorator.onQuery(span, statement, queries);
+
+        info.addCustomValue(SPAN, span);
+    }
+
+    @Override
+    @SneakyThrows(SQLException.class)
+    public void afterQuery(final ExecutionInfo info, final List<QueryInfo> queries) {
+        final Span span = info.getCustomValue(SPAN, Span.class);
+
+        if (!info.isSuccess()) {
+            final Statement statement = info.getStatement();
+            final Throwable error = info.getThrowable();
+
+            decorator.onError(span, statement, error);
```

Comment:
Is it an issue that in case the decorator fails (i.e. exception thrown in `onError`) that we never gonna call `span.finish()`?

Reply:
True. Good catch.

Reply:
Now I'm curious how vendors/implementations deal with that. Technically it could be the source of a memory leak to start but never finish spans. Ideally tracers would keep only weak references to spans and discard them.

Reply:
I added a `try-finally` around `onError` but technically the `onQuery` could fail also and what to do with the span then? Discard (if, how)? Ignore and finish as usual?

Reply:
If I'm reading LightStep's code correctly then they don't keep a strong reference to it and abandoned spans would be garbage collected.

Reply:
Question regarding errors during `onQuery` still stands though.

Reply:
If `onQuery` fails, is `afterQuery` (and hence `onError`) called?

Reply:
https://github.com/ttddyy/datasource-proxy/blob/c1f62baf6fa13014982e8098fc3a22ffb592892d/src/main/java/net/ttddyy/dsproxy/proxy/StatementProxyLogic.java#L307-L396

`beforeQuery` is running outside the `try-finally` that would execute the `afterQuery`.